### ISSUE TITLE
Plugin's with option type 'float' or 'complex' fail

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#258]
In GitLab by @mschwager on Sep 20, 2018, 14:34

*Please describe how you installed Flake8*

```
$ pip install flake8
```

**Note**: Some *nix distributions patch Flake8 arbitrarily to accommodate incompatible software versions. If you're on one of those distributions, your issue may be closed and you will be asked to open an issue with your distribution package maintainers instead.

*Please provide the exact, unmodified output of `flake8 --bug-report`*

```
$ flake8 --bug-report
{
  "dependencies": [
    {
      "dependency": "setuptools",
      "version": "39.2.0"
    }
  ],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.6.5",
    "system": "Linux"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "cohesion",
      "version": "0.8.0"
    },
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.3.1"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "1.6.0"
    }
  ],
  "version": "3.5.0"
}
```

*Please describe the problem or feature*

I wrote a Flake8 plugin that provides its own arguments. The problem is noticed, and well described, here: https://github.com/mschwager/cohesion/issues/9.

After digging deeper, it looks like Flake8 uses `optparse` on the backend. However, when specifying `"type": "float"` like my code does here:

```
kwargs = {
    ...
    'type': 'float',
    ...
}
config_opts = getattr(parser, 'config_options', None)
if isinstance(config_opts, list):
    # flake8 2.x
    ...
else:
    # flake8 3.x
    parser.add_option(flag, **kwargs)
```

It doesn't actually consider the argument received from `setup.cfg` as a `float`. I confirmed this by setting a breakpoint at the code in question and confirming that it was a `str`:

```
/tmp/test # PYTHONPATH=/tmp/cohesion/lib:$PYTHONPATH flake8 --enable-extension='H' test.py 
> /tmp/cohesion/lib/cohesion/flake8_extension.py(41)parse_options()
     40         import ipdb; ipdb.set_trace()
---> 41         cls.cohesion_below = options.cohesion_below
     42 

ipdb> options.cohesion_below
'100.0'
ipdb> type(options.cohesion_below)
<class 'str'>

```

Now, I'm not entirely sure why, but it looks like Flake8 has its own shim for type conversion (added here https://gitlab.com/pycqa/flake8/commit/ae794fb46a177c754071c7a36811893c8dedc710) in `options.manager.Option.normalize_from_setuptools` (shouldn't `optparse` be doing this for us?), but it doesn't support `float` or `complex`. I see a few options for fixing this:

*  Investigate why `optparse` isn't handling these conversions for us.
*  Update `normalize_from_setuptools` to handle all of `optparse`'s [standard-option-types](https://docs.python.org/3.6/library/optparse.html#standard-option-types)

The second option would probably look something like:

```
    def normalize_from_setuptools(self, value):
        """Normalize the value received from setuptools."""
        value = self.normalize(value)
        if self.type == 'int' or self.action == 'count':
            return int(value)
        elif self.type == 'float':
            return float(value)
        elif self.type == 'complex':
            return complex(value)
        if self.action in ('store_true', 'store_false'):
            value = str(value).upper()
            if value in ('1', 'T', 'TRUE', 'ON'):
                return True
            if value in ('0', 'F', 'FALSE', 'OFF'):
                return False
        return value
```

*If this is a bug report, please explain with examples (and example code) what you expected to happen and what actually happened.*

I expected my plugin option `options.cohesion_below` to be a `float` not a `str`.

If nobody else picks this up I'll probably just open a merge request with the changes suggested in the second option above.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Oct 20, 2018, 09:57

I'd happily merge a PR that implements the second option. The first isn't feasible because it requires digging into internal functionality of optparse (which is why `normalize_from_setuptools` exists).

**Comment by asottile** [comment]
In GitLab by @ymdatta on Oct 28, 2018, 05:08

mentioned in merge request !261

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jan 17, 2019, 05:50

closed via commit 0f3f4ff9b4e7901695484ee1ea1fbcbc19632149

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jan 17, 2019, 05:50

closed via merge request !261

**Comment by asottile** [comment]
In GitLab by @ymdatta on Jan 17, 2019, 05:50

closed via commit 6341d1382e0ab14043396e8a9fbb18c27a3e236c
